In [2]:
import pandas as pd
import numpy as np

In [3]:
hospital= pd.read_csv("D:\Datascience\Grey Sloan Hospital Data.csv")

In [4]:
hospital

male  age  education  currentSmoker  cigsPerDay  BPMeds  \
0        1   39        4.0              0         0.0     0.0   
1        0   46        2.0              0         0.0     0.0   
2        1   48        1.0              1        20.0     0.0   
3        0   61        3.0              1        30.0     0.0   
4        0   46        3.0              1        23.0     0.0   
...    ...  ...        ...            ...         ...     ...   
4230     1   51        3.0              1        43.0     0.0   
4231     0   48        2.0              1        20.0     0.0   
4232     0   52        2.0              0         0.0     0.0   
4233     1   40        3.0              0         0.0     0.0   
4234     0   39        3.0              1        30.0     0.0   

      prevalentStroke  prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  \
0                   0             0         0    195.0  106.0   70.0  26.97   
1                   0             0         0    250.0  121.0   81.0  28.73   
2                   0             0         0    245.0  127.5   80.0  25.34   
3                   0             1         0    225.0  150.0   95.0  28.58   
4                   0             0         0    285.0  130.0   84.0  23.10   
...               ...           ...       ...      ...    ...    ...    ...   
4230                0             0         0    207.0  126.5   80.0  19.71   
4231                0             0         0    248.0  131.0   72.0  22.00   
4232                0             0         0    269.0  133.5   83.0  21.47   
4233                0             1         0    185.0  141.0   98.0  25.60   
4234                0             0         0    196.0  133.0   86.0  20.91   

      heartRate  glucose  CHD chance  
0          80.0     77.0           0  
1          95.0     76.0           0  
2          75.0     70.0           0  
3          65.0    103.0           1  
4          85.0     85.0           0  
...         ...      ...         ...  
4230       65.0     68.0           0  
4231       84.0     86.0           0  
4232       80.0    107.0           0  
4233       67.0     72.0           0  
4234       85.0     80.0           0  

[4235 rows x 16 columns]

In [5]:
hospital.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4235 entries, 0 to 4234
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   male             4235 non-null   int64  
 1   age              4235 non-null   int64  
 2   education        4131 non-null   float64
 3   currentSmoker    4235 non-null   int64  
 4   cigsPerDay       4206 non-null   float64
 5   BPMeds           4183 non-null   float64
 6   prevalentStroke  4235 non-null   int64  
 7   prevalentHyp     4235 non-null   int64  
 8   diabetes         4235 non-null   int64  
 9   totChol          4185 non-null   float64
 10  sysBP            4235 non-null   float64
 11  diaBP            4235 non-null   float64
 12  BMI              4216 non-null   float64
 13  heartRate        4234 non-null   float64
 14  glucose          3853 non-null   float64
 15  CHD chance       4235 non-null   int64  
dtypes: float64(9), int64(7)
memory usage: 529.5 KB


In [6]:
hospital.isnull().sum()

male                 0
age                  0
education          104
currentSmoker        0
cigsPerDay          29
BPMeds              52
prevalentStroke      0
prevalentHyp         0
diabetes             0
totChol             50
sysBP                0
diaBP                0
BMI                 19
heartRate            1
glucose            382
CHD chance           0
dtype: int64

In [7]:
hospital.skew()

male                0.283907
age                 0.228269
education           0.688641
currentSmoker       0.025040
cigsPerDay          1.247203
BPMeds              5.548558
prevalentStroke    12.904414
prevalentHyp        0.819360
diabetes            5.992083
totChol             0.871733
sysBP               1.145580
diaBP               0.713960
BMI                 0.982732
heartRate           0.645224
glucose             6.215122
CHD chance          1.938578
dtype: float64

In [8]:
hospital['education'].fillna(hospital['education'].mean(),inplace=True)

In [9]:
hospital['cigsPerDay'].fillna(hospital['cigsPerDay'].median(),inplace=True)

In [10]:
hospital['BPMeds'].fillna(hospital['BPMeds'].median(),inplace=True)

In [11]:
hospital['totChol'].fillna(hospital['totChol'].mean(),inplace=True)

In [12]:
hospital['BMI'].fillna(hospital['BMI'].mean(),inplace=True)

In [13]:
hospital['heartRate'].fillna(hospital['heartRate'].mean(),inplace=True)

In [14]:
hospital['glucose'].fillna(hospital['glucose'].mean(),inplace=True)

In [15]:
hospital.isnull().sum()

male               0
age                0
education          0
currentSmoker      0
cigsPerDay         0
BPMeds             0
prevalentStroke    0
prevalentHyp       0
diabetes           0
totChol            0
sysBP              0
diaBP              0
BMI                0
heartRate          0
glucose            0
CHD chance         0
dtype: int64

In [17]:
x=hospital.drop(columns=['CHD chance'])

In [18]:
x

male  age  education  currentSmoker  cigsPerDay  BPMeds  \
0        1   39        4.0              0         0.0     0.0   
1        0   46        2.0              0         0.0     0.0   
2        1   48        1.0              1        20.0     0.0   
3        0   61        3.0              1        30.0     0.0   
4        0   46        3.0              1        23.0     0.0   
...    ...  ...        ...            ...         ...     ...   
4230     1   51        3.0              1        43.0     0.0   
4231     0   48        2.0              1        20.0     0.0   
4232     0   52        2.0              0         0.0     0.0   
4233     1   40        3.0              0         0.0     0.0   
4234     0   39        3.0              1        30.0     0.0   

      prevalentStroke  prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  \
0                   0             0         0    195.0  106.0   70.0  26.97   
1                   0             0         0    250.0  121.0   81.0  28.73   
2                   0             0         0    245.0  127.5   80.0  25.34   
3                   0             1         0    225.0  150.0   95.0  28.58   
4                   0             0         0    285.0  130.0   84.0  23.10   
...               ...           ...       ...      ...    ...    ...    ...   
4230                0             0         0    207.0  126.5   80.0  19.71   
4231                0             0         0    248.0  131.0   72.0  22.00   
4232                0             0         0    269.0  133.5   83.0  21.47   
4233                0             1         0    185.0  141.0   98.0  25.60   
4234                0             0         0    196.0  133.0   86.0  20.91   

      heartRate  glucose  
0          80.0     77.0  
1          95.0     76.0  
2          75.0     70.0  
3          65.0    103.0  
4          85.0     85.0  
...         ...      ...  
4230       65.0     68.0  
4231       84.0     86.0  
4232       80.0    107.0  
4233       67.0     72.0  
4234       85.0     80.0  

[4235 rows x 15 columns]

In [19]:
y=hospital[['CHD chance']]

In [20]:
y

CHD chance
0              0
1              0
2              0
3              1
4              0
...          ...
4230           0
4231           0
4232           0
4233           0
4234           0

[4235 rows x 1 columns]

In [21]:
from sklearn.model_selection import train_test_split

In [22]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.3,random_state=42)

In [23]:
len(x_train),len(x_test),len(y_train),len(y_test)

(2964, 1271, 2964, 1271)

In [24]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

In [25]:
lr_hosp = LogisticRegression()
model_lr_hosp = lr_hosp.fit(x_train,y_train)

C:\Users\Mrunmayee\anaconda3\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
C:\Users\Mrunmayee\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [26]:
dt_hosp= DecisionTreeClassifier()
model_dt_hosp = dt_hosp.fit(x_train,y_train)

In [27]:
rf_hosp = RandomForestClassifier(n_estimators=200,random_state=42)
model_rf = rf_hosp.fit(x_train,y_train)

<ipython-input-27-67a42a7c1db5>:2: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  model_rf = rf_hosp.fit(x_train,y_train)


In [28]:
y_test['Predict_LR'] = model_lr_hosp.predict(x_test)

<ipython-input-28-343fb75159b2>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  y_test['Predict_LR'] = model_lr_hosp.predict(x_test)


In [29]:
y_test['Predict_DT'] = model_dt_hosp.predict(x_test)

<ipython-input-29-22fd901bf2fe>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  y_test['Predict_DT'] = model_dt_hosp.predict(x_test)


In [30]:
y_test['Predict_RF'] = model_rf.predict(x_test)

<ipython-input-30-efc987ea5d9c>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  y_test['Predict_RF'] = model_rf.predict(x_test)


In [31]:
from sklearn.metrics import classification_report

In [32]:
print(classification_report(y_test['CHD chance'],y_test['Predict_LR']))

              precision    recall  f1-score   support

           0       0.86      1.00      0.92      1078
           1       0.71      0.06      0.11       193

    accuracy                           0.85      1271
   macro avg       0.78      0.53      0.52      1271
weighted avg       0.83      0.85      0.80      1271



In [33]:
print(classification_report(y_test['CHD chance'],y_test['Predict_DT']))

              precision    recall  f1-score   support

           0       0.86      0.83      0.84      1078
           1       0.21      0.25      0.23       193

    accuracy                           0.74      1271
   macro avg       0.53      0.54      0.53      1271
weighted avg       0.76      0.74      0.75      1271



In [34]:
print(classification_report(y_test['CHD chance'],y_test['Predict_RF']))

              precision    recall  f1-score   support

           0       0.86      0.99      0.92      1078
           1       0.54      0.07      0.12       193

    accuracy                           0.85      1271
   macro avg       0.70      0.53      0.52      1271
weighted avg       0.81      0.85      0.80      1271



## We can conclude that accuracy is good for logistic regression and random forest which is 85% for both. Followed by 74% for decision tree. So it is clear that logistic regression and random forest models  are good for this case study.